discovery/linode/linode.go

```diff
@@ -0,0 +1,261 @@
+func (c *SDConfig) SetDirectory(dir string) {
+	c.HTTPClientConfig.SetDirectory(dir)
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	return c.HTTPClientConfig.Validate()
+}
+
+// Discovery periodically performs Linode requests. It implements
+// the Discoverer interface.
+type Discovery struct {
+	*refresh.Discovery
+	client       *linodego.Client
+	port         int
+	tagSeparator string
+}
+
+// NewDiscovery returns a new Discovery which periodically refreshes its targets.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	d := &Discovery{
+		port:         conf.Port,
+		tagSeparator: conf.TagSeparator,
+	}
+
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "linode_sd", config.WithHTTP2Disabled())
+	if err != nil {
+		return nil, err
+	}
+
+	client := linodego.NewClient(
+		&http.Client{
+			Transport: rt,
+			Timeout:   time.Duration(conf.RefreshInterval),
+		},
+	)
+	client.SetUserAgent(fmt.Sprintf("Prometheus/%s", version.Version))
+	d.client = &client
+
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"linode",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	tg := &targetgroup.Group{
+		Source: "Linode",
+	}
+
+	// Gather all linode instances.
+	instances, err := d.client.ListInstances(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// Gather detailed IP address info for all IPs on all linode instances.
+	detailedIPs, err := d.client.ListIPAddresses(ctx, &linodego.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, instance := range instances {
+		if len(instance.IPv4) == 0 {
+			continue
+		}
+
+		var (
+			privateIPv4, publicIPv4, publicIPv6             string
+			privateIPv4RDNS, publicIPv4RDNS, publicIPv6RDNS string
+			backupsStatus                                   string
+			extraIPs                                        []string
+		)
+
+		for _, ip := range instance.IPv4 {
+			for _, detailedIP := range detailedIPs {
+				if detailedIP.Address != ip.String() {
+					continue
+				}
+
+				if detailedIP.Public && publicIPv4 == "" {
+					publicIPv4 = detailedIP.Address
+
+					if detailedIP.RDNS != "" && detailedIP.RDNS != "null" {
+						publicIPv4RDNS = detailedIP.RDNS
+					}
+				} else if !detailedIP.Public && privateIPv4 == "" {
+					privateIPv4 = detailedIP.Address
+
+					if detailedIP.RDNS != "" && detailedIP.RDNS != "null" {
+						privateIPv4RDNS = detailedIP.RDNS
+					}
+				} else {
+					extraIPs = append(extraIPs, detailedIP.Address)
+				}
+			}
+		}
+
+		if instance.IPv6 != "" {
+			for _, detailedIP := range detailedIPs {
+				if detailedIP.Address != strings.TrimSuffix(instance.IPv6, "/128") {
```

Comment:
This seems fragile. Can we use the netmask in detailedIP ?
detailedIP.Address + "/" + detailedIP.Netmask != instance.IPv6

Reply:
Duh, silly me, that's a much better idea. I'll update. 👍

Reply:
Agreed that this is definitely fragile, I'll give it some more thought. 

The info for IPv6 addresses returned in the networking endpoint returns the IPs as part of a `/64`, while the address is actually a `/128` and is shown as such in the IPv6 info in the instances response, so I'm not quite sure the best way to compare them. 

Would it be better to check + split `instances.IPv6` on `/` and only take the first part? Rather than hardcode the subnet to remove from the end of the string?